### ISSUE TITLE
So, is coc.vim better than LanguageClient.vim?

### ISSUE THREAD

**Issue by AdamWagner** [hallettj/dot-vim#1]
I'm beginning to explore Intellisense-esque functionality via LSPs. The landscape is rather confusing at the moment! Since it seems like you've tried several approaches, it'd be great to get your review.

**Comment by hallettj** [comment]
Hi, sure I can write a bit about my experience. The short answer is that yes, in my opinion coc.nvim is the best solution that I have tried. My main point of reference is Typescript, because that is what I have been mostly working with since I installed coc.nvim. I also do a lot of work with Flow, and I do a little bit of Rust programming, and I have been using the corresponding language servers.

[LanguageClient-neovim](https://github.com/autozimu/LanguageClient-neovim) works quite nicely for LSP integration. But because it is generalized it does not add any language-specific enhancements. One issue that I had is that I wanted to restart the Typescript server from time to time, and I did not see any way to do that through LanguageClient. It could be that it is difficult to generalize that behavior across language server implementations.

I used [nvim-typescript](https://github.com/mhartington/nvim-typescript) for a while to get Typescript-specific enhancements, while continuing to use LanguageClient for Flow and Rust. Nvim-typescript includes commands for restarting the Typescript server. I also found that it has additional perks, such as a command to automatically add an import line for the symbol under the cursor. That is a real time saver!

After reading some recommendations on Reddit I decided to try [coc.nvim](https://github.com/neoclide/coc.nvim). I am impressed! Instead of relying exclusively on LSP coc manages language-specific add-ons that can provide their own commands, and provide features such as code actions. With coc I get the command to restart the Typescript server, and the feature to automatically add import lines. Even better, thanks to the popupmenu integration when I confirm a completion the relevant import is added without any explicit action from me. I also get a number of code actions which can do things like extract selected code into new functions. Coc is also good at presenting type definitions and errors where I can see them. Relevant type information appears in a preview window; but on top of that using autocomplete on functions triggers a snippet expansion that essentially pastes the function parameter list into the buffer.

Something to consider is that coc replaces deoplete. I have used deoplete for quite a while, and it has worked very nicely for me. The documentation lists some reasons why coc's completion handling is better; what I have specifically noticed is that the ability to trigger relevant actions automatically when confirming a completion is quite helpful (for example adding to the imports list), and the fuzzy matching feature is nice. Coc's handling of the popupmenu is more opinionated than deoplete's: I don't remember seeing this mentioned in the documentation, but coc works best when you use these complete options:

```vim
set completeopt=noinsert,noselect,menuone
```

I wrote about how those settings affect my workflow in [an issue](https://github.com/neoclide/coc.nvim/issues/262).

Deoplete has the advantage that many plugins will automatically populate its completions list. I don't know if coc will work the same way. But coc also replaces the plugins that I was using completions from, so losing that integration is not a problem for me.

The fact that coc delegates to third-party extensions means that I am managing a second set of plugins using a new plugin system. I think that is an acceptable cost because I get a lot of useful features in exchange. And writing coc extensions should be simpler than writing custom Neovim plugins for each language because coc extensions can rely on coc for common functionality.

For languages that do not have an existing coc extension, but that do have a language server, it is possible to configure basic LSP integration without writing an extension. For example I have this in my `~/.config/nvim/coc-settings.json` configuration to set up LSP integration with Flow:

```json
"languageserver": {
  "flow": {
    "command": "flow-language-server",
    "args": ["--stdio", "--try-flow-bin"],
    "filetypes": ["javascript", "javascript.jsx"],
    "initializationOptions": {},
    "settings": {},
    "rootPatterns": [".flowconfig"]
  }
},
"tsserver.enableJavascript": false
```

The last line disables Typescript integration for plain Javascript files so that I can use Flow instead for those files.
